File: ruby/lib/resp_bench/command/impl/set_command.rb
# frozen_string_literal: true

module RespBench
  module Command
    module Impl
      # SET command implementation.
      class SetCommand < Command
        def initialize(config)
          super
          # Pre-generate value data of the configured size
          @value = generate_value(@data_size_bytes)
        end

        def execute(client, key_generator)
          key = key_generator.next_key
          result = client.set(key, @value)
          CommandResult.new(
            command_name: @name,
            latency_micros: result.latency_micros,
            success: result.success?
          )
        end

        private

        def generate_value(size)
          # Generate a deterministic value of the specified size
          # Using a simple pattern for reproducibility
          pattern = "0123456789ABCDEF"
          (pattern * ((size / pattern.length) + 1))[0, size]
        end
      end
    end
  end
end

File: python/src/resp_bench/command/impl/set_command.py
"""SET command with a pre-generated deterministic value."""

from __future__ import annotations

from typing import Optional

from ...client.benchmark_client import AsyncBenchmarkClient
from ...config.command_config import CommandConfig
from ..command import Command, CommandResult

_PATTERN = b"0123456789ABCDEF"


class SetCommand(Command):
    def __init__(self, config: CommandConfig) -> None:
        super().__init__(config)
        self._value = self._generate_value(self._data_size_bytes)

    async def execute(self, client: AsyncBenchmarkClient, key: Optional[str]) -> CommandResult:
        result = await client.set(key, self._value)
        return CommandResult(
            command_name=self.name,
            latency_micros=result.latency_micros,
            success=result.success,
        )

    @staticmethod
    def _generate_value(size: int) -> bytes:
        repeats = (size // len(_PATTERN)) + 1
        return (_PATTERN * repeats)[:size]
